### PR TITLE
feature: add MixLoRA model support, improve inference speed and framework design

### PR DESCRIPTION
**Pull request by mikecovlee**
## Features

- Add MixLoRA Model Support
  - [MixLoRA](https://github.com/TUDB-Labs/multi-lora-fine-tune/blob/mixlora_support/MixLoRA.md) is a LoRA based Mix-of-Expert adapter model
  - Support two different routing strategies: Mixtral and Switch Transformers (`mlora/MixLoRA.py`)
- Improve Inference Speed
  - Support kv-cache and top-p sampling
  - Rewrite Web UI of inference based on m-LoRA (`inference.py`)
- Improve Framework Design
  - Generic MoE support (`mlora/FeedForward.py`)
  - More user-friendly APIs (see example in `generate.py`)
  - Uniform logging mechanism
  - Build-in trainer and generator (`mlora/train.py` and `mlora/generate.py`)
 
## TODO

- [ ] Fix Web UI
- [x] ChatGLM support
- [ ] ChatGLM inference
- [ ] Evaluations for MixLoRA
- [ ] Integrate evaluation methods into m-LoRA

### CODE REVIEW

File: MixLoRA.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# MixLoRA: Resource-Efficient Model with Mix-of-Experts Architecture for Enhanced LoRA Performance
+
+<div align="left"><img src="./assets/Mix-LoRA.png" width=60%"></div>
+
+The fundamental concept of MixLoRA is based on a pre-trained model with all parameters frozen, such as LLaMA-7B. It involves training multiple LoRA expert modules on top of its fully connected layer (FFN). Simultaneously, a routing layer (Gate Linear) is trained, creating a more powerful Mixture of Experts (MoE) language model. Theoretically, this approach allows achieving performance similar to existing MoE models but with fewer resources.
+
```
Comment (merlintang) [inline MixLoRA.md:6]: tell why this happen?

**Comment by yezhem** [comment]
1. move the Prompter to prompter.py
2. prompter do not use hard code
3. del the utils.py

File: mlora/modelargs.py [hunk 2]
```diff
@@ -49,3 +49,107 @@ class MultiLoraBatchData:
...
+
+
+@dataclass
+class LoraConfig:
+    adapter_name_: str = ""
+    device_: str = "cuda:0"
+    lora_r_: int = None
+    lora_alpha_: int = None
+    lora_dropout_: float = None
+    target_modules_: dict = None
```
Comment (yezhem) [inline mlora/modelargs.py:61]: dict -> Dict

File: mlora/modelargs.py [hunk 3]
```diff
@@ -49,3 +49,107 @@ class MultiLoraBatchData:
...
+
+    def init(self, config: dict) -> "LoraConfig":
+        self.lora_r_ = config["r"]
```
Comment (yezhem) [inline mlora/modelargs.py:64]: check the config param

File: mlora/modelargs.py [hunk 4]
```diff
@@ -49,3 +49,107 @@ class MultiLoraBatchData:
...
+@dataclass
+class LoraConfig:
+    adapter_name_: str = ""
+    device_: str = "cuda:0"
+    lora_r_: int = None
+    lora_alpha_: int = None
+    lora_dropout_: float = None
+    target_modules_: dict = None
+
+    def init(self, config: dict) -> "LoraConfig":
```
Comment (yezhem) [inline mlora/modelargs.py:63]: init -> from_configxxx

**Review by yezhem — CHANGES_REQUESTED** [review]

File: MixLoRA.md (hunk 1 above)
Comment (mikecovlee) [reply MixLoRA.md:42]: docs updated

**Review by yezhem — APPROVED** [review]
LGTM

**Pull request merged**